src/Utils/obiDialogParser.test.ts

```diff
@@ -0,0 +1,202 @@
+import * as Util from './util'
+import * as fs from 'fs-extra'
+import * as fspath from 'path'
+import * as klaw from 'klaw-sync'
+
+describe('obiDialogParser', () => {
+    /**
+     * Given a path __relative to this test file__, returns an absolute path to the same file.
+     */
+    function absolutePathFromPathRelativeToCurrentFile(inputPath: string): string {
+        const thisFilePath = module.filename
+        const thisFileDir = fspath.dirname(fspath.resolve(thisFilePath))
+        return fspath.join(thisFileDir, inputPath)
+    }
+
+    /**
+     * Given an absolute path or path relative to CWD of test execution, returns a `File` object wrapping that file.
+     */
+    function pathToFileObject(path: string): File {
+        const content = fs.readFileSync(path)
+        return new File([content], path)
+    }
+
+    describe('Dialog reconstruction test', () => {
+        /**
+         * This tests the reconstruction of a ProductKey dialog that has some complex properties,
+         * such as `SwitchCondition` nodes, `ScorerSteps` that need to be appended to previous
+         * `TrainDialogRounds`.
+         */
+        test('Test product key example', async () => {
+            // TODO(thpar) : investigate why this test fails on CircleCI.
+            const envOS: string | undefined = process.env.OS
+            if (!envOS || !envOS.toLocaleLowerCase().startsWith("win")) {
+                console.log("Skipping Dialog reconstruction test due to incompatible OS type")
+                return
+            }
+
+            // Collect array of input files for the OBI dialog parser.
+            const path = absolutePathFromPathRelativeToCurrentFile('../__testdata__/product_key_dialogs')
+            const globResults = klaw(path)
+            const dialogFiles: File[] = []
+            for (const pathItem of globResults) {
+                if (!pathItem.stats.isFile()) {
+                    // Skip non-file objects.
+                    continue
+                }
+                dialogFiles.push(pathToFileObject(pathItem.path))
+            }
+            const actions: CLM.ActionBase[] = []
+            const fakeCreateActionThunk = (appId: string, action: CLM.ActionBase) => {
+                return new Promise<CLM.ActionBase>((resolve) => {
+                    // Update the action id and return the action.
+                    let outputAction = Util.deepCopy(action)
+                    outputAction.actionId = `${actions.length}`
+                    actions.push(outputAction)
+                    resolve(outputAction)
+                })
+            }
+            const entities: CLM.EntityBase[] = []
+            const fakeCreateEntityThunk = (appId: string, entity: CLM.EntityBase) => {
+                return new Promise<CLM.EntityBase>((resolve) => {
+                    // Update the entity id and return the entity.
+                    let outputEntity = Util.deepCopy(entity)
```

Comment:
Didn't notice this before. I've been trying to get people to follow convention of using `const` wherever possible so when `let` is used you know it was needed and the variable will be reassigned.  Similar to usage of `Map`s. I wasn't sure if you knew or not.

Reply:
Yep, still doing that mental switch in my head from C++ to TS.  `const` in C++ doesn't let you mutate the variable.  Updated.